<a href="https://colab.research.google.com/github/le-incroyable1-dev/tinker_with_Tensorflow/blob/main/Convolutions_and_Pooling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [12]:
import tensorflow as tf
from tensorflow import keras

In [13]:
fmnist = tf.keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fmnist.load_data()

In [15]:
model = keras.Sequential([
    keras.layers.Flatten(),
    keras.layers.Dense(128, activation = tf.nn.relu),
    keras.layers.Dense(10, activation = tf.nn.softmax)
])

model.compile(optimizer = 'adam', loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'])

model.fit(train_images, train_labels, epochs = 5)

test_loss = model.evaluate(test_images, test_labels)
print(test_loss)

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 4.1214 - accuracy: 0.7090
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.6840 - accuracy: 0.7543
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.5912 - accuracy: 0.7942
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.5476 - accuracy: 0.8134
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.5555 - accuracy: 0.8173
[0.555490255355835, 0.817300021648407]


In [8]:
convolutedModel = keras.Sequential([
    keras.layers.Conv2D(64, (3,3), activation = 'relu', input_shape = (28, 28, 1)),
    keras.layers.MaxPooling2D(2, 2),
    keras.layers.Conv2D(64, (3,3), activation = 'relu'),
    keras.layers.MaxPooling2D(2, 2),
    keras.layers.Flatten(input_shape = (28, 28)),
    keras.layers.Dense(128, activation = tf.nn.relu),
    keras.layers.Dense(10, activation = tf.nn.softmax)
])

Refer : https://lodev.org/cgtutor/filtering.html for more information on convolution filters

In [9]:
convolutedModel.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_4 (Conv2D)           (None, 26, 26, 64)        640       
                                                                 
 max_pooling2d_4 (MaxPooling  (None, 13, 13, 64)       0         
 2D)                                                             
                                                                 
 conv2d_5 (Conv2D)           (None, 11, 11, 64)        36928     
                                                                 
 max_pooling2d_5 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 flatten_4 (Flatten)         (None, 1600)              0         
                                                                 
 dense_8 (Dense)             (None, 128)              

In [16]:
convolutedModel.compile(optimizer = 'adam', loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'])

convolutedModel.fit(train_images, train_labels, epochs = 5)
test_loss = convolutedModel.evaluate(test_images, test_labels)
print(test_loss)

Epoch 1/5
1875/1875 [==============================] - 98s 52ms/step - loss: 0.5712 - accuracy: 0.8219
Epoch 2/5
1875/1875 [==============================] - 95s 51ms/step - loss: 0.3351 - accuracy: 0.8769
Epoch 3/5
1875/1875 [==============================] - 97s 52ms/step - loss: 0.2955 - accuracy: 0.8894
Epoch 4/5
1875/1875 [==============================] - 96s 51ms/step - loss: 0.2687 - accuracy: 0.8988
Epoch 5/5
313/313 [==============================] - 5s 15ms/step - loss: 0.3213 - accuracy: 0.8882
[0.3212888538837433, 0.8881999850273132]


Network accuracy increased to 0.88 and loss reduced to 0.3123